torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -686,6 +819,109 @@ SegmentedGroup* SegmentCandidateFinder::mergeNodes() {
   return last_merged;
 }
 
+// Logic largely parallels mergeNodes, but they are used
+//  in different phases of segmentation. Should consider
+//  a clean up and share the implementations.
```

Comment:
What would prevent you from either calling mergeNodes multiples times, or calling this instead of mergeNodes with only 2 groups?

Reply:
I guess not much beyond trying to limit complexity and amount of change in one PR. 

`calling mergeNodes multiples times`:
It can be done but I'd rather not merge nodes pair by pair. In most cases where `mergeAllGivenGroups` is used, if we try to perform the merge pair-by-pair, some calls to `mergeNodes` will introduce cycles to `SegmentedFusion` and rely on subsequent `mergeNodes` to fix it. I'd appreciate not seeing this behavior unless by design.

`calling this instead of mergeNodes with only 2 groups`:
I'd prefer this idea to the previous one except this is mostly refactor. Currently `mergeNodes` doesn't exactly take 2 arguments but merges a `to_merge` list pair-by-pair. There would be some refactor work to do which I'd prefer to do in a separate PR.